lib/upipe-modules/upipe_row_split.c

```diff
@@ -0,0 +1,357 @@
+    upipe_row_split_init_ubuf_mgr(upipe);
+    upipe_row_split_init_output(upipe);
+
+    return upipe;
+}
+
+static void upipe_row_split_input(struct upipe *upipe, struct uref *uref,
+                                  struct upump **upump_p)
+{
+    if (!upipe_row_split_check_input(upipe)) {
+        upipe_row_split_hold_input(upipe, uref);
+        upipe_row_split_block_input(upipe, upump_p);
+    } else if (!upipe_row_split_handle(upipe, uref, upump_p)) {
+        upipe_row_split_hold_input(upipe, uref);
+        upipe_row_split_block_input(upipe, upump_p);
+        /* Increment upipe refcount to avoid disappearing before all packets
+         * have been sent. */
+        upipe_use(upipe);
+    }
+}
+
+static bool upipe_row_split_handle(struct upipe *upipe, struct uref *uref,
+                                  struct upump **upump_p)
+{
+    struct upipe_row_split *upipe_row_split = upipe_row_split_from_upipe(upipe);
+    if (!upipe_row_split->ubuf_mgr)
+        return false;
+
+    size_t hsize, vsize;
+    uint8_t macropixel;
+    if (unlikely(!ubase_check(uref_pic_size(uref, &hsize, &vsize,
+                        &macropixel)))) {
+        upipe_warn(upipe, "dropping picture");
+        upipe_throw_error(upipe, UBASE_ERR_INVALID);
+        uref_free(uref);
+        return true;
+    }
+
+    uint64_t vsize_slice = upipe_row_split->slice_height;
+    if (vsize_slice > vsize)
+        vsize_slice = vsize;
+
+    for (uint64_t done = 0; done < vsize; done += vsize_slice) {
+        struct ubuf *ubuf = ubuf_pic_alloc(upipe_row_split->ubuf_mgr, hsize, vsize_slice);
+        if (!ubuf)
+            abort();
+
+        const char *chroma;
+        ubuf_pic_foreach_plane(ubuf, chroma) {
+            const uint8_t *src;
+            uint8_t *dst;
+            if (!ubase_check(ubuf_pic_plane_write(ubuf, chroma, 0, 0, -1, -1, &dst))) {
+                abort();
+            }
+
+            if (done + vsize_slice > vsize) {
+                vsize_slice = vsize - done;
+                ubuf_pic_resize(ubuf, 0, 0, hsize, vsize_slice);
+                /* TODO: do we care if this fails? */
+            }
+
+            if (!ubase_check(uref_pic_plane_read(uref, chroma, 0, done, -1, vsize_slice, &src))) {
+                abort();
+            }
+
+            size_t dst_stride;
+            uint8_t dst_hsub, dst_vsub, dst_macropixel_size;
+            if (unlikely(!ubase_check(ubuf_pic_plane_size(ubuf, chroma, &dst_stride,
+                        &dst_hsub, &dst_vsub, &dst_macropixel_size)))) {
+                abort();
+            }
+
+            memcpy(dst, src, vsize_slice * dst_stride);
+            // TODO: dst stride != src stride
+
+            ubase_assert(uref_pic_plane_unmap(uref, chroma, 0, done, -1, vsize_slice));
+            ubase_assert(ubuf_pic_plane_unmap(ubuf, chroma, 0, 0, -1, -1));
+
+        }
+
+        struct uref *uref_slice = uref_fork(uref, ubuf);
+
+        uref_attr_set_unsigned(uref_slice,
```

Comment:
Please use the macro UREF_ATTR_UNSIGNED in header.

Reply:
Ideally these would be used by more modules in the future, not sure in which common location I should put them

Reply:
uref_pic ?